### ISSUE TITLE
Errors in logging can result in log observers not being restored

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4379]
|<img alt="catlee's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| catlee reported|
|-|-|
|Trac ID|trac#4379|
|Type|defect|
|Created|2010-03-23 19:21:43Z|
|Branch|https://github.com/twisted/twisted/tree/logging-errors-4379|

One particular example is when a RuntimeError is hit.  When twisted tries to log the fact that the observer failed to log the message, another RuntimeError is generated, which results in the original observer not being restored.

Attachments:

* [twisted-logging.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ed/3ed54b86efcc05e5ca25e668628027cc585e911a/e831b861b64b550614a971878c23b687a627ffe3.patch) (4175 bytes) - added by catlee on 2010-03-23 19:24:08Z - 
* [4379.bugfix](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ed/3ed54b86efcc05e5ca25e668628027cc585e911a/ba23692fdd426ec79622f4b5e6b1c4e3ba6afaac.bugfix) (126 bytes) - added by catlee on 2010-03-26 13:59:10Z - news file entry

<details><summary>Searchable metadata</summary>

```
trac-id__4379 4379
type__defect defect
reporter__catlee catlee
priority__normal normal
milestone__ 
branch__branches_logging_errors_4379 branches/logging-errors-4379
branch_author__catlee__itamarst catlee, itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1269372103000000 1269372103000000
changetime__1308920202000000 1308920202000000
version__None None
owner__itamar itamar
cc__thijs cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra removed owner|
|-|-|

Thanks Chris. Could you also supply a NEWS file, as described on TwistedDevelopment? Also adding the review keyword so it moves into that queue.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

Replying to [thijs](#note_1):
> Thanks Chris. Could you also supply a NEWS file, as described on TwistedDevelopment?

Sorry that's actually [ReviewProcess#Newsfiles this page].

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to catlee|
|-|-|

Thanks.  A few mechanical things:

  1. `test_runtimeError` should have a docstring
  1. `assertTrue(x >= y)` is preferred over `assert_(x >= y)`
  1. `assertIn(x, y)` is preferred over `assert_(x in y)`
  1. `increasedRuntimeLimited` should be documented and, very probably, private.

Perhaps more interestingly:

  5. `test_runtimeError` fails when I run it.  `self.assert_("RuntimeError" in i['message'][0])` doesn't succeed.  Here are all the messages in catcher: [{'message': ('At the bottom of everything (939)',), 'time': 1271006674.191, 'isError': 0, 'system': '-'}, {'message': ('Back on top',), 'time': 1271006674.191169, 'isError': 0, 'system': '-'}]
  1. I wonder if the test would be simpler if instead of calling a global `err` function, `twisted.python.log.LogPublisher.msg` could call `err` on itself (and in turn `err` could then use `self.msg` instead of a global `msg`).  This would seem to remove the need for patching the log module.  If you feel like it, this might make a good separate change (ie, new ticket), since it's generally useful and not really tied to fixing this bug.  If not, that's fine too.  I might take it on. :)
  1. Is the `flushLoggedErrors` call supposed to actually flush anything?  It's not for me (that is, it returns `[]`).

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

[#4407](https://github.com/twisted/twisted/issues/4407) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [32081]) Branching to 'logging-errors-4379'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

(In [32082]) Patch from catlee. Refs [#4379](https://github.com/twisted/twisted/issues/4379)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst removed owner|
|-|-|

Ready for review (unless something fails here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/logging-errors-4379).

I switched to a more general solution that covers cases other than blowing the stack, and a test that doesn't rely on blowing the stack. I did not address review comment #6, since it sounded optional.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @itamarst|
|-|-|

 * There are a couple of assertEquals which should be replaced by assertEqual.

Please merge once fixed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set status to `closed`|
|-|-|

(In [32167]) Merge logging-errors-4379.
Fixes: [#4379](https://github.com/twisted/twisted/issues/4379)
Author: catlee, itamar
Reviewer: therve

If logging to log observer fails, and logging this resulting failure also fails, don't break future logging.
